[AArch64][SVE] Add partial reduction SDNodes

JamesChesterman opened the pull request:
Add the opcode 'ISD::PARTIAL_REDUCE_ADD' and use it when making SDNodes. When the inputs and outputs have types that can allow for lowering to wide add or dot product instruction(s), then convert the corresponding intrinsic to an SDNode. This will allow legalisation, which will be added in a future patch, to be done more easily.

JamesChesterman committed:
[AArch64][SVE] Add partial reduction SDNodes

Add the opcode 'ISD::PARTIAL_REDUCE_ADD' and use it when making
SDNodes. When the inputs and outputs have types that can allow for
lowering to wide add or dot product instruction(s), then convert
the corresponding intrinsic to an SDNode. This will allow
legalisation, which will be added in a future patch, to be done
more easily.

MacDue commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1451,6 +1451,11 @@ enum NodeType {
+  // The `llvm.experimental.vector.partial.reduce.add` intrinsic
+  // Operands: Accumulator, Input
+  // Outputs: Output
+  PARTIAL_REDUCE_ADD,
```
Should this be `EXPERIMENTAL_PARTIAL_REDUCE_ADD` (like `EXPERIMENTAL_VECTOR_HISTOGRAM`)?

MacDue commented on llvm/include/llvm/CodeGen/SelectionDAG.h:
```diff
@@ -1595,6 +1595,11 @@ class SelectionDAG {
   /// the target's desired shift amount type.
   SDValue getShiftAmountOperand(EVT LHSTy, SDValue Op);
 
+  /// Get a partial reduction SD node for the DAG. This is done when the input
+  /// and output types can be legalised for wide add(s) or dot product(s)
+  SDValue getPartialReduceAddSDNode(SDLoc DL, SDValue Chain, SDValue Acc,
+                                    SDValue Input);
+
   /// Create the DAG equivalent of vector_partial_reduce where Op1 and Op2 are
   /// its operands and ReducedTY is the intrinsic's return type.
   SDValue getPartialReduceAdd(SDLoc DL, EVT ReducedTy, SDValue Op1,
```
I don't think there should be `DAG.getPartialReduceAddSDNode()` and `DAG.getPartialReduceAdd()`. 

I think `DAG.getPartialReduceAdd()` should always return the ISD node, and then in `SelectionDAGLegalize::ExpandNode(SDNode *Node)` in `LegalizeDAG.cpp`, the ISD node can be expanded (how it currently is in `DAG.getPartialReduceAdd()`) for targets that don't have a more specified lowering. 

Note: In `TargetLoweringBase::initActions()` you should probably set the default legalization action for `PARTIAL_REDUCE_ADD` to `Expand`. Then in  `AArch64ISelLowering` I think you'll want to set the action to `Custom` for supported types and handle the lowering to dsot/wide adds there.

JamesChesterman replied:
Changed the name of `getPartialReduceAdd` to `expandPartialReduceAdd`. For now, I'll try to carry on using the DAG combines, although I may need to take the approach described in your comment when legalisation is implemented.

MacDue commented on llvm/include/llvm/CodeGen/SelectionDAGNodes.h:
```diff
@@ -3010,6 +3010,22 @@ class MaskedHistogramSDNode : public MaskedGatherScatterSDNode {
   }
 };
 
+class PartialReduceAddSDNode : public SDNode {
+public:
+  friend class SelectionDAG;
```
A special class is probably not needed (most nodes don't have these), and I think there may be some issues (see next comment).

JamesChesterman replied:
Done

MacDue commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -2452,6 +2452,12 @@ SDValue SelectionDAG::getShiftAmountOperand(EVT LHSTy, SDValue Op) {
 
+SDValue SelectionDAG::getPartialReduceAddSDNode(SDLoc DL, SDValue Chain,
+                                                SDValue Acc, SDValue Input) {
+  return getNode(ISD::PARTIAL_REDUCE_ADD, DL, Acc.getValueType(), Chain, Acc,
+                 Input);
```
I think this is just constructing a plain `SDNode *` (not a `PartialReduceAddSDNode *`), which I think would technically make the `auto *PR = cast<PartialReduceAddSDNode>(N)` later on UB (might fail with `-DLLVM_USE_SANITIZER=Undefined`), even though the layout of both classes is the same.

I think to use your custom class would be: `newSDNode<PartialReduceAddSDNode>(...)`.

JamesChesterman replied:
Removed this function. `DAG.getNode` is instead used wherever this function was called.

JamesChesterman committed:
Changes to previous patch. Involves removing PartialReduceAddSDNode
as well as changing how the intrinsic is transformed into the SD
node.

MacDue commented on llvm/include/llvm/CodeGen/SelectionDAG.h:
```diff
@@ -1595,10 +1595,10 @@ class SelectionDAG {
-  /// its operands and ReducedTY is the intrinsic's return type.
-  SDValue getPartialReduceAdd(SDLoc DL, EVT ReducedTy, SDValue Op1,
-                              SDValue Op2);
+  /// Expands partial reduce node which can't be lowered to wide add or dot
+  /// product instruction(s)
```
SelectionDAG.h is not target specific, so I don't think this should mention the wide add/dot product lowering (which is only for AArch64). Also, could you document the operands (ReducedTY, Op1, Op2)?

MacDue commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1451,6 +1451,11 @@ enum NodeType {
   VECREDUCE_UMAX,
   VECREDUCE_UMIN,
 
+  // The `llvm.experimental.vector.partial.reduce.add` intrinsic
```
nit:
```suggestion
  // This corresponds to the `llvm.experimental.vector.partial.reduce.add` intrinsic
```

JamesChesterman committed:
Remove unnecessary function parameter and update comments

JamesChesterman committed:
Code formatting changes

MacDue commented on llvm/include/llvm/CodeGen/SelectionDAG.h:
```diff
@@ -1595,10 +1595,11 @@ class SelectionDAG {
-  SDValue getPartialReduceAdd(SDLoc DL, EVT ReducedTy, SDValue Op1,
-                              SDValue Op2);
+  /// Expands PARTIAL_REDUCE_ADD nodes which can't be lowered.
+  /// @param Op1 Accumulator for where the result is stored for the partial
+  /// reduction operation
+  /// @param Op2 Input for the partial reduction operation
```
nit: `\p` is more commonly used to mark parameters.

JamesChesterman replied:
Done

MacDue commented on llvm/include/llvm/CodeGen/SelectionDAG.h:
```diff
@@ -1595,10 +1595,11 @@ class SelectionDAG {
-  /// its operands and ReducedTY is the intrinsic's return type.
-  SDValue getPartialReduceAdd(SDLoc DL, EVT ReducedTy, SDValue Op1,
-                              SDValue Op2);
+  /// Expands PARTIAL_REDUCE_ADD nodes which can't be lowered.
```
nit: I don't think "which can't be lowered" is relevant to this function, I think it's better to say what the function does. Maybe something "Expands PARTIAL_REDUCE_ADD nodes to a sequence of subvector extracts followed by vector adds"

JamesChesterman replied:
Done

paulwalker-arm commented:
I've not had chance to look at the PR yet but from a high level I think having an ISD node that matches the definition of the intrinsic this closely is going to make legalisation and selection harder than necessary.  I was expecting the ISD node(s) to be signed operations because then you open the possibility of the operands having different element types which will make it much easier to legalise efficiently as well as providing a route to potentially isel directly.

Specifically, I'm suggesting you follow a similar idiom as used by other nodes where implicitly extension is beneficial (e.g. MULH, ABD) and implement `PARTIAL_VECREDUCE_SADD` and `PARTIAL_VECREDUCE_UADD`.

JamesChesterman committed:
Make two ISD nodes for partial reductions as opposed to one

JamesChesterman commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -462,6 +462,16 @@ class TargetLoweringBase {
     return true;
   }
 
+  /// Return true if there is a sign extend on the input to this function. Used
+  /// to determine whether to transform the
+  /// @llvm.experimental.vector.partial.reduce.* intrinsic to
+  /// PARTIAL_REDUCE_SADD or PARTIAL_REDUCE_UADD. It also removes the extend
+  /// from the input. \p Input The 'Input' operand to the
+  /// @llvm.experimental.vector.partial.reduce.* intrinsic.
+  virtual bool isPartialReductionInputSigned(SDValue &Input) const {
+    return false;
+  }
```
I just thought I'd do it this way in case other targets had a slightly different implementation of partial reductions? For example, originally I was also going to get it to remove the extends preceding dot products (before realising that I'd need more ISD nodes for sudot and usdot). But I thought I'd leave it like this in case other targets wanted to place other operations between the extend and the partial reduction?

JamesChesterman replied:
Also, in the future legalisation patch, I use it to see whether there is an `ISD::MUL` operation done on the input, which leads to different behaviour being done. This searches for the dot product IR pattern, which is target-specific.

JamesChesterman replied:
Removed this function, explained in `SelectionDAGBuilder.cpp` comment.

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8126,15 +8126,21 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
 
     if (!TLI.shouldExpandPartialReductionIntrinsic(cast<IntrinsicInst>(&I))) {
-      visitTargetIntrinsic(I, Intrinsic);
+      if (TLI.isPartialReductionInputSigned(Input))
+        setValue(&I,
+                 DAG.getNode(ISD::PARTIAL_REDUCE_SADD, dl, AccVT, Acc, Input));
+      else
+        setValue(&I,
+                 DAG.getNode(ISD::PARTIAL_REDUCE_UADD, dl, AccVT, Acc, Input));
```
The signedness is a property of the operation rather than a a property of its data.  This is why the partial.reduce.add intrinsic does not require signedness information because it requires all operands to be the same size and thus there is no implicit extension happening as part of the operation.

During initial selection the choice of `ISD::PARTIAL_REDUCE_SADD` to `ISD::PARTIAL_REDUCE_UADD` does not matter because it will also have all operands being the same size.  In such circumstances you should just pick one as the canonical form.

The choice of node only becomes important when a transformation changes the operation to take a second operand that has a smaller element size than the first.  This is because it is converting an explicit extension into an implicit one.  Doing this is most likely going to be the result of a DAG combine or perhaps legalisation depending on the circumstance.

JamesChesterman replied:
Changed this so that only the `shouldExpandPartialReductionIntrinsic` function is called, and the node outputted is always `ISD::PARTIAL_REDUCE_UADD`. Searching for exts and removing them is now done in the DAG combine. I decided to do it here rather than in legalisation so that, if the exts are different, `usdot` instructions can be outputted here.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -21808,45 +21822,40 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
-             Intrinsic::experimental_vector_partial_reduce_add &&
-         "Expected a partial reduction node");
-
   if (!Subtarget->isSVEorStreamingSVEAvailable())
```
This existing code looks wrong to me because the wide add instructions are only available in SVE2 (or SME in streaming mode).  I'm pretty sure this code will lead to an internal compiler error when only SVE is available and is worth fixing asap as a new PR.

JamesChesterman replied:
Done

JamesChesterman committed:
Determine which ISD node to use in DAG combine rather than in
SelectionDAGBuilder.

JamesChesterman commented on llvm/include/llvm/CodeGen/SelectionDAG.h:
```diff
@@ -1595,10 +1595,12 @@ class SelectionDAG {
+  /// followed by vector adds.
+  /// \p Op1 Accumulator for where the result is stored for the partial
+  /// reduction operation.
+  /// \p Op2 Input for the partial reduction operation.
```
Done

JamesChesterman commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -21834,15 +21824,23 @@ SDValue tryLowerPartialReductionToWideAdd(SDNode *N,
+    if (InputOpcode == ISD::SIGN_EXTEND)
+      Opcode = ISD::PARTIAL_REDUCE_SADD;
+  }
+
```
Currently there would be no node for `ISD::PARTIAL_REDUCE_SADD` made anymore. But with the legalisation, I would make it output a node with the opcode `ISD::PARTIAL_REDUCE_SADD` here. This would then go through legalisation and the execution would then lead back to `tryLowerPartialReductionToWideAdd`. 
I originally left this out of this patch because it wasn't necessary to get it to work. But I will go back and make it return the node, just to make sure it all works properly.

JamesChesterman replied:
I've now made it so the DAG-combine function decides whether to keep the node opcode as `ISD::PARTIAL_REDUCE_UADD` or whether to convert it to `ISD::PARTIAL_REDUCE_SADD`. But, the DAG-combine function also includes the lowering for `usdot` IR patterns (due to the reasons described in the above comment). Also, the DAG-combine needed to include lowering for `(nx)v16i8` to `(nx)v4i64`, because otherwise the instruction would be attempted to be legalised. But, there is now a lowering function called by `LowerOperation` which covers the general cases for lowering to wide add / dot product instructions.

JamesChesterman committed:
Separate lowering code for PARTIAL_REDUCE_U/SADD

Separate lowering code from all being in the DAG-combine function.
Now the DAG-combine decides whether the node should be the signed
or unsigned version of partial reduce add. Then there is a function
in LowerOperation that does the actual lowering to wide adds or dot
products if it is able to.

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1839,8 +1841,17 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
+    }
   }
 
+  for (auto VT : {MVT::v4i64, MVT::v4i32, MVT::v2i32}) {
+    setOperationAction(ISD::PARTIAL_REDUCE_UADD, VT, Custom);
+    setOperationAction(ISD::PARTIAL_REDUCE_SADD, VT, Custom);
+  }
```
Which cases are these fixed-vector types supported? (Neon dot products?) Your custom lowering will probably need to check if the vectors are fixed/scalable as otherwise there may be cases it attempts to make fixed-size wide adds? E.g acc = v2i32, input = v4i32?

JamesChesterman replied:
Done

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -28916,6 +28925,39 @@ SDValue AArch64TargetLowering::LowerVECTOR_HISTOGRAM(SDValue Op,
+
+  if (AccVT.getVectorElementCount() * 4 == InputVT.getVectorElementCount()) {
+    unsigned IndexAdd = 0;
+    // ISD::MUL may have already been lowered, meaning the operands would be in
+    // different positions.
+    if (Input.getOpcode() != ISD::MUL)
+      IndexAdd = 1;
+    auto A = Input.getOperand(IndexAdd);
```
This looks unsafe. If the `ISD::MUL` is not there then you can't assume it's a dot product?

JamesChesterman replied:
In lowering, the `ISD::MUL` is lowered before the partial reduction node. So I've now made it so if the `Input` opcode is NOT `ISD::MUL` and `Input`'s number of elements is 4 times greater than `Acc`'s number of elements, then it is expanded. This now happens in the DAG-combine.

JamesChesterman committed:
Change the way the dot product pattern is checked for lowering.
Add condition in wide add combine to not allow fixed length
vectors.

paulwalker-arm commented:
Perhaps a left field suggestion but looking at the use cases you want to handle I'm wondering if my initial suggestion was too limiting and instead we'd be better of with something more powerful like `PARTIAL_REDUCE_SMLA/PARTIAL_REDUCE_UMLA`.  That way you'll have an easier time representing the expected use cases of the intrinsic whilst still being able to represent the most general form by using a unit vector for one of the multiplicands.

This will re-raise an earlier question of yours as to whether we need `PARTIAL_REDUCE_USMLA`. To that I would personally hold off and maintain the current DAG combine because it feels somewhat target specific, but at the same time I wouldn't object to it if the consensus is in favour of it.

JamesChesterman committed:
Change from adding ISD::PARTIAL_REDUCE_S/UADD to adding
ISD::PARTIAL_REDUCE_S/UMLA

This makes the lowering function easier as you do not need to worry
about whether the MUL is lowered or not. Instead its operands are
taken from it. If there is no MUL instruction and just one operand,
the other operand is a vector of ones (for value types eligible for
wide add lowering).

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8147,7 +8147,7 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
     if (!TLI.shouldExpandPartialReductionIntrinsic(cast<IntrinsicInst>(&I))) {
       setValue(&I,
-               DAG.getNode(ISD::PARTIAL_REDUCE_UADD, dl, AccVT, Acc, Input));
+               DAG.getNode(ISD::PARTIAL_REDUCE_UMLA, dl, AccVT, Acc, Input));
```
Correct. For this initial building of the DAG I recommend ignoring the operand analysis and just always create the extra vector of ones operand.  Then you can implement a DAG combine to transform the `PARTIAL_REDUCE_UMLA` based on the operands (i.e. remove an explicit multiply and/or extensions).

MacDue replied:
It looks like you've left this to `tryCombineToDotProduct`, but that means the node you create here is broken (since it's missing an operand, and is not meant to multiply anything).

JamesChesterman replied:
Done, made `MUL` instructions get generated all the time in the combine and expand functions. So if `Input2` is a splat vector of all 1s, the instruction is automatically removed. I have tested that this works by trying it with a splat vector of all 3s, and the `MUL` instructions are generated and remain in the output.

JamesChesterman committed:
MUL instructions now included in DAG combines.

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -29164,6 +29188,32 @@ SDValue AArch64TargetLowering::LowerVECTOR_HISTOGRAM(SDValue Op,
+
+  unsigned Opcode = Op.getOpcode();
+
+  if (AccVT.getVectorElementCount() * 4 == InputVT.getVectorElementCount()) {
+    unsigned DotOpcode = Opcode == ISD::PARTIAL_REDUCE_SMLA ? AArch64ISD::SDOT
+                                                            : AArch64ISD::UDOT;
+    return DAG.getNode(DotOpcode, DL, AccVT, Acc, Input1, Input2);
+  }
+  bool InputIsSigned = Opcode == ISD::PARTIAL_REDUCE_SMLA;
+  unsigned BottomOpcode =
+      InputIsSigned ? AArch64ISD::SADDWB : AArch64ISD::UADDWB;
+  unsigned TopOpcode = InputIsSigned ? AArch64ISD::SADDWT : AArch64ISD::UADDWT;
+  auto BottomNode = DAG.getNode(BottomOpcode, DL, AccVT, Acc, Input1);
+  return DAG.getNode(TopOpcode, DL, AccVT, BottomNode, Input1);
```
I don't think this should assume `Input2` is a vector of all ones if `AccVT.getVectorElementCount() * 4 != InputVT.getVectorElementCount()`.

JamesChesterman replied:
I've now made it so a `MUL` instruction is made here if `Input2` is not a splat vector of constant 1s. A `MUL` instruction is still made in the DAG combine because then it can be removed if an operand is a vector of constant 1s, or can be made into a shift if an operand is a power of 2. This would not happen if a `MUL` instruction was just made in `LowerPARTIAL_REDUCE_MLA`. In the DAG combine, Input2 is set to a vector of constant 1s, so the `MUL` is not repeated.

JamesChesterman committed:
Make the no bin op changes work with adding Partial Reduction
SDNodes.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8135,15 +8135,19 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
-                                         getValue(I.getOperand(0)),
-                                         getValue(I.getOperand(1))));
+    setValue(&I,
+             DAG.expandPartialReduceAdd(
```
Did you consider always expanding the intrinsic to a PARTIAL_REDUCE_UMLA/SMLA ISD node, and only expand the operation in LegalizeVectorOps.cpp when there is no custom-lowering for it?

That avoids the need for calling `DAG.expandPartialReduceAdd` in `tryCombineToDotProduct`.

JamesChesterman replied:
Thank you for the recommendation. I've done it this way now, so the intrinsic is always changed to the ISD node in `SelectionDAGBuilder.cpp`. This way I've been able to remove `shouldExpandPartialReductionIntrinsic()` too. I've just put a call to `DAG.expandPartialReduceAdd` at the end of `performPartialReduceAddCombine()` in `AArch64ISelLowering.cpp`.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -2046,11 +2057,18 @@ bool AArch64TargetLowering::shouldExpandPartialReductionIntrinsic(
-  if (Op1VT.getVectorElementType() == VT.getVectorElementType() &&
-      (VT.getVectorElementCount() * 4 == Op1VT.getVectorElementCount() ||
-       VT.getVectorElementCount() * 2 == Op1VT.getVectorElementCount()))
+  auto Input = I->getOperand(1);
+  EVT InputVT = EVT::getEVT(Input->getType());
+
+  if ((InputVT == MVT::nxv4i64 && VT == MVT::nxv2i64) ||
+      (InputVT == MVT::nxv8i32 && VT == MVT::nxv4i32) ||
+      (InputVT == MVT::nxv16i16 && VT == MVT::nxv8i16) ||
+      (InputVT == MVT::nxv16i64 && VT == MVT::nxv4i64) ||
+      (InputVT == MVT::nxv16i32 && VT == MVT::nxv4i32) ||
+      (InputVT == MVT::nxv8i64 && VT == MVT::nxv2i64) ||
+      (InputVT == MVT::v16i64 && VT == MVT::v4i64) ||
+      (InputVT == MVT::v16i32 && VT == MVT::v4i32) ||
+      (InputVT == MVT::v8i32 && VT == MVT::v2i32))
```
Is there a reason for rewriting it like this? (did the original condition allow a case that it shouldn't have?)

JamesChesterman replied:
The removed code would allow types that could be used but would just need to go through legalisation first. Originally, and on the current upstream main branch, at the end of the DAG combine it would return `DAG.expandPartialReduceAdd` (it's `DAG.getPartialReduceAdd` on upstream main). This meant that any types that weren't legal would just end up going to this function. However, when changing the end of the DAG combine to return `SDValue()` instead, these cases would try to go to legalisation, where currently there is not a case for partial reduction ISD nodes.

For now, I've just changed it back to returning the expand function at the end of the DAG combine. When I add legalisation, I'll do the type checking during the DAG combine instead of here.

sdesmalen-arm reviewed (commented):
There are no changes to any of the tests, is that expected?

JamesChesterman committed:
Address comments on patch. Remove shouldExpandPartialReductionIntrinsic().

JamesChesterman commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1451,6 +1451,13 @@ enum NodeType {
+  // Nodes used to represent a partial reduction addition operation (signed and
+  // unsigned).
+  // Operands: Accumulator, Input
+  // Outputs: Output
```
Done, specified what needs to be passed into the node as its arguments. It will be similar to what is specified for the intrinsic itself, so `%a` and`%b` need to have the same element type as `%acc`.

JamesChesterman commented on llvm/include/llvm/CodeGen/SelectionDAG.h:
```diff
@@ -1602,10 +1602,17 @@ class SelectionDAG {
+  /// \p Op1 Accumulator for where the result is stored for the partial
+  /// reduction operation.
+  /// \p Op2 Input for the partial reduction operation.
+  /// Expands PARTIAL_REDUCE_S/UMLA nodes.
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8135,15 +8135,19 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
 
     if (!TLI.shouldExpandPartialReductionIntrinsic(cast<IntrinsicInst>(&I))) {
-      visitTargetIntrinsic(I, Intrinsic);
+      setValue(&I, DAG.getNode(ISD::PARTIAL_REDUCE_UMLA, dl, AccVT, Acc, Input,
```
Yes this is arbitrarily chosen. The node is then changed to its correct signedness in the DAG combine. Added a comment explaining this and asserts for the operand types.

JamesChesterman commented:
> There are no changes to any of the tests, is that expected?

Yes this is expected. This patch is to change around the design of the ISD node to make it easier to use now as well as making it easier to use in future scenarios.

JamesChesterman committed:
Add the MUL in LowerPARTIAL_REDUCE_MLA()

Only do it if Input2 is a splat vector of constant 1s. Still create
the MUL in the DAG combine for the wide add pattern. This is
because it is pruned if an operand is constant 1s, or changed to
a shift instruction if an operand is a power of 2. This would not
happen if the MUL was made in LowerPARTIAL_REDUCE_MLA.

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -22011,34 +22010,25 @@ static SDValue tryCombineWhileLo(SDNode *N,
+    MulOpLHSIsSigned = MulOpRHSIsSigned = (Op1Opcode == ISD::SIGN_EXTEND);
+    MulOpLHS = Op1->getOperand(0);
+    MulOpRHS = DAG.getAnyExtOrTrunc(Op2, DL, MulOpLHS.getValueType());
+  } else if (Op1Opcode == ISD::MUL) {
+    SDValue ExtMulOpLHS = Op1->getOperand(0);
+    SDValue ExtMulOpRHS = Op1->getOperand(1);
```
I'm a little confused by what this fold is doing now. `PARTIAL_REDUCE_MLA` is `acc += partial_reduce(op1 * op2)`. But this combine is also checking for another `MUL` node. I think this would be simpler to follow if there was a fold that took:


```
mul = op1 * op2
ret = PARTIAL_REDUCE_MLA(mul * splat (1))
```

And folded it to:

```
ret = PARTIAL_REDUCE_MLA(op1, op2)
```

Then the dot product lowering can be simpler (since it does not need to worry about the `MUL` node).

JamesChesterman replied:
Done. Now there is a separate function for lowering:
`PARTIAL_REDUCE_MLA (Acc, MUL(EXT(MulOpLHS), EXT(MulOpRHS)), SPLAT (1))`
To:
`PARTIAL_REDUCE_MLA(Acc, EXT(MulOpLHS), EXT(MulOpRHS))`
The function `tryCombineToDotProduct` can then handle this pattern (and removes the extends) as well as the pattern for `PARTIAL_REDUCE_MLA(Acc, EXT(Op), SPLAT (1))`.
I've added comments in the relevant places detailing what happens.

MacDue commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -2467,19 +2467,24 @@ SDValue SelectionDAG::getShiftAmountOperand(EVT LHSTy, SDValue Op) {
-SDValue SelectionDAG::getPartialReduceAdd(SDLoc DL, EVT ReducedTy, SDValue Op1,
-                                          SDValue Op2) {
-  EVT FullTy = Op2.getValueType();
+SDValue SelectionDAG::expandPartialReduceAdd(SDLoc DL, SDValue Acc,
```
nit:
```suggestion
SDValue SelectionDAG::expandPartialReduceMLA(SDLoc DL, SDValue Acc,
```

MacDue commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -22072,77 +22067,102 @@ SDValue tryLowerPartialReductionToDot(SDNode *N,
-      !(ExtOpVT == MVT::nxv8i16 && AccVT == MVT::nxv4i32) &&
-      !(ExtOpVT == MVT::nxv16i8 && AccVT == MVT::nxv8i16))
-    return SDValue();
+SDValue performPartialReduceAddCombine(SDNode *N, SelectionDAG &DAG,
```
nit:
```suggestion
SDValue performPartialReduceMLACombine(SDNode *N, SelectionDAG &DAG,
```

JamesChesterman committed:
Separate the DAG combine into two stages for when there is a Mul
operation as Input1.

Also change the LangRef in ISDOpcodes.h for PARTIAL_REDUCE_MLA
nodes to set restrictions on what can be used for its inputs.
Rename functions to accord to PARTIAL_REDUCE_MLA rather than
PARTIAL_REDUCE_ADD.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -22011,138 +22010,188 @@ static SDValue tryCombineWhileLo(SDNode *N,
-  if (!Scalable && (!Subtarget->isNeonAvailable() || !Subtarget->hasDotProd()))
+
+  SDValue ExtMulOpLHS = Op1->getOperand(0);
+  SDValue ExtMulOpRHS = Op1->getOperand(1);
+  unsigned ExtMulOpLHSOpcode = ExtMulOpLHS->getOpcode();
+  unsigned ExtMulOpRHSOpcode = ExtMulOpRHS->getOpcode();
+  if (!ISD::isExtOpcode(ExtMulOpLHSOpcode) ||
+      !ISD::isExtOpcode(ExtMulOpRHSOpcode))
     return SDValue();
 
-  SDLoc DL(N);
+  SDValue MulOpLHS = ExtMulOpLHS->getOperand(0);
+  SDValue MulOpRHS = ExtMulOpRHS->getOperand(0);
+  EVT MulOpLHSVT = MulOpLHS.getValueType();
+  if (MulOpLHSVT != MulOpRHS.getValueType())
+    return SDValue();
```
With the way you've currently written this function, my understanding is that this code can be removed, because this code does not try to fold any of the extends into the PARTIAL_REDUCE operation. i.e. al it is doing is:
```
PARTIAL_REDUCE_UMLA(acc, MUL(lhs, rhs), SPLAT(1))
->
PARTIAL_REDUCE_UMLA(acc, lhs, rhs)
```

sdesmalen-arm commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1451,6 +1451,21 @@ enum NodeType {
   VECREDUCE_UMIN,
 
+  // Partial Reduction nodes. These represent multiply-add instructions because
+  // Input1 and Input2 are multiplied together first. This result is then
```
I thought the behaviour of this node is that Input1 and Input2 are first sign- (for SMLA) or zero- (for UMLA) extended to be the same element type as Acc.

nit: `Input1` and `Input2` are not defined yet. What about starting with a description of the node with operands, e.g. `PARTIAL_REDUCE_*MLA(Acc, Input1, Input2)` ?

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -22011,138 +22010,188 @@ static SDValue tryCombineWhileLo(SDNode *N,
+  if (Op0ElemVT != Op1ElemVT || Op1.getOpcode() == AArch64ISD::USDOT)
+    return SDValue(N, 0);
+
+  if (auto MLA = tryCombinePartialReduceMLAMulOp(Op0, Op1, Op2, DAG, DL)) {
```
This seems like a generically useful combine, so it can be moved to the generic combine code in `DAGCombiner.cpp`

JamesChesterman commented:
This work is being split into separate Pull Requests and is being done differently.
See here for first PR:
https://github.com/llvm/llvm-project/pull/125207
